gson/src/main/java/com/google/gson/stream/JsonReader.java

```diff
@@ -737,7 +737,7 @@ private int peekNumber() throws IOException {
     }
 
     // We've read a complete number. Decide if it's a PEEKED_LONG or a PEEKED_NUMBER.
-    if (last == NUMBER_CHAR_DIGIT && fitsInLong && (value != Long.MIN_VALUE || negative) && (value!=0 || false==negative)) {
+    if (last == NUMBER_CHAR_DIGIT && fitsInLong && (value != Long.MIN_VALUE || negative) && (value!=0 || !negative)) {
```

Comment:
The negation could also be moved outside the parentheses and it might then be worth adding a comment explaining why this check exists:
```suggestion
    if (last == NUMBER_CHAR_DIGIT && fitsInLong && (value != Long.MIN_VALUE || negative)
        // Don't lose information about -0, user might want to read it as double -0.0
        && !(value == 0 && negative)
    ) {
```
(not sure if this exactly follows the style guide)

Reply:
`false==negative` is certainly a bit eccentric. I agree that we should at least write `!negative`.

I think this complicated boolean expression can be simplified slightly:
```java
last == NUMBER_CHAR_DIGIT && fitsInLong && (negative ? value != 0 : value != Long.MIN_VALUE)
```
My reasoning is that if `negative` is true, `(value != Long.MIN_VALUE || negative)` is also true, and if `negative` is false, `(value != 0 || !negative)` is true. So we only need to check the other one of those two conditions in each case. We could also refactor further, to `value != (negative ? 0 : Long.MIN_VALUE)`, but I'm not sure that's an improvement.

Reply:
> I think this complicated boolean expression can be simplified slightly:
> 
> ```java
> last == NUMBER_CHAR_DIGIT && fitsInLong && (negative ? value != 0 : value != Long.MIN_VALUE)
> ```

But a comment to explain this condition would be good nonetheless, maybe something like this?
```
// Don't store -0 as long; user might want to read it as double -0.0
// Don't try to convert Long.MIN_VALUE to positive long; it would overflow MAX_VALUE
```

Reply:
Agreed. Those comments look good.